src/library/scala/collection/immutable/Set.scala

```diff
@@ -21,8 +21,11 @@ import scala.language.higherKinds
 
 
 /** Base trait for immutable set collections */
-trait Set[A] extends Iterable[A] with collection.Set[A] with SetOps[A, Set, Set[A]] {
-  override def iterableFactory: IterableFactory[IterableCC] = Set
+trait Set[A] extends Iterable[A]
+    with collection.Set[A]
+    with SetOps[A, Set, Set[A]]
+    with IterableFactoryDefaults[A, Set] {
```

Comment:
Why does `Set` extend `IterableFactoryDefault`? Intuitively, this pattern should either be usable everywhere (i.e. `Iterable` extends `IterableFactoryDefault`) or only in concrete classes. I suspect it's the former, so effectively this PR moves the methods that depend on the protected CC aliases into their own traits, which makes the compiler recheck the types in subclasses.

Reply:
I was also just looking at this in the context of documenting, we should give it another try.

Reply:
```
[error] /Users/adriaan/git/scala/src/library/scala/collection/immutable/Map.scala:150:73: type mismatch;
[error]  found   : Set[K] (in scala.collection) 
[error]  required: Set[K] (in scala.collection.immutable) 
[error]     def incl(elem: K): Set[K] = if (this(elem)) this else empty ++ this + elem
[error]                                                                         ^
[error] /Users/adriaan/git/scala/src/library/scala/collection/immutable/Map.scala:151:63: type mismatch;
[error]  found   : Set[K] (in scala.collection) 
[error]  required: Set[K] (in scala.collection.immutable) 
[error]     def excl(elem: K): Set[K] = if (this(elem)) empty ++ this - elem else this
[error]                                                               ^
[error] two errors found
```